src/display/annotation_layer.js

```diff
@@ -159,6 +161,7 @@ class AnnotationElement {
     this.annotationStorage = parameters.annotationStorage;
     this.enableScripting = parameters.enableScripting;
     this.hasJSActions = parameters.hasJSActions;
+    this._fieldObjects = parameters.fieldObjects;
```

Comment:
nit: I think that it'd be slightly better to have a `Map` `name => [ids....]` generated from `parameters.fieldObjects` where fields with `page === -1` are removed this way we don't have to check in `_getElementsByName`. Since we needn't other properties than id, it should help to slightly reduce memory use.

Reply:
> [...] with `page === -1` are removed this way we don't have to check in `_getElementsByName`.

You're correct that this would slightly reduce the necessary parsing here, however...

> it should help to slightly reduce memory use.

That would actually *increase* overall memory usage, since it'd force us to allocate one `Map` per page whereas currently every page simply gets *the same* Object (since the API implements caching).

---

Do you have any pointers for the integration-test, e.g. more specifically what you'd like to see tested?
(I've not written any integration-tests before, so feeling slightly out of my depth here.)

Reply:
> > [...] with `page === -1` are removed this way we don't have to check in `_getElementsByName`.
> 
> You're correct that this would slightly reduce the necessary parsing here, however...
> 
> > it should help to slightly reduce memory use.
> 
> That would actually _increase_ overall memory usage, since it'd force us to allocate one `Map` per page whereas currently every page simply gets _the same_ Object (since the API implements caching).
> 
I always forget that there is a layer per page...

> Do you have any pointers for the integration-test, e.g. more specifically what you'd like to see tested?
> (I've not written any integration-tests before, so feeling slightly out of my depth here.)
I'd imagine a pdf with on page 1: a text field, a radio and a checkbox and the same on page 2 and page 4 (or more if needed) with the same fields with the same name.
Fill elements on page 1 then scroll on page 2, check that the values are correct (they should have been updated thanks to document.get...) and the same with page 4 (updated through annotationStorage).

This test could help (after reading it, I fixed the current bug but only when the field is updated from JS... :( )
https://github.com/mozilla/pdf.js/blob/master/test/integration/scripting_spec.js#L730
The line here:
https://github.com/mozilla/pdf.js/blob/master/test/integration/scripting_spec.js#L723
open the given pdf and the form is ready when the element with the given id is displayed.
and here you've a way to scroll to the page 4:
https://github.com/mozilla/pdf.js/blob/master/test/integration/scripting_spec.js#L770

Reply:
Maybe we could create the simplified fieldObjects before to pass it to the layer.

Reply:
@Snuffleupagus, if you don't have time to write a test or don't feel like to do it, tell me and I'll write it later today or tomorrow.

Reply:
> Maybe we could create the simplified fieldObjects before to pass it to the layer.

We could, but I don't want to do that for two reasons:
 - It would make the viewer components more difficult to use standalone, since it'd then not be enough just to call the correct API-method but you'd also need to *manually* transform the data into the required format.
 - It's not inconceivable that we'll want to add some future `fieldObjects` usage in the AnnotationLayer, and if so we'd then need yet *another* external transformation or a breaking change of e.g. the `DefaultAnnotationLayerFactory` class.

All-in-all, I'm going to leave this as-is since the very slight overhead during parsing still ought to be lower than having to use `document.getElementsByName` unconditionally as in the current code.

---

> if you don't have time to write a test or don't feel like to do it, tell me and I'll write it later today or tomorrow.

I'm guessing that it would be *a lot quicker* for you to add a test, and I'd probably won't have time to dive into all of that until the weekend.
So if you don't mind writing the integration-test, I'd be very grateful for your help :-)

Reply:
@Snuffleupagus I'll add a test tomorrow.